add catOptions and equalsWith functions for seq, list, array and map

sideeffffect opened the pull request.

forki commented on src/FSharpx.Collections/Collections.fs:
```diff
@@ -230,6 +230,13 @@ module Seq =
 
+    /// The catOptions function takes a list of Options and returns a seq of all the Some values.
+    let catOptions (xs:seq<Option<'a>>) =
+        seq {
```
where is the difference to `Seq.choose id`?

sideeffffect replied:
none, the point is to have one function that does this, so that you don't have to compose two other to do the job (like in Haskell [catMaybes](http://hackage.haskell.org/package/base/docs/Data-Maybe.html#v:catMaybes))

forki replied:
(Seq.choose id) is also one function and can be composed!?

sideeffffect replied:
yes, but that's already two variables that have to have parentheses around them -- `catOptions` is supposed to be only a small helper function, I wrote it because I've found myself writing `(Seq.choose id)` way to often. [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)

sideeffffect replied:
I can rewrite them in terms of `(choose id)` if you think that would be better

forki replied:
mmm tbh I don't really agree gere. But maybe we find others that can join discussion

sideeffffect commented:
I've implemented the `catOptions`s using `choose id` and inlined them -- I hope that makes them acceptable now
I've also added `equalsWith` and helpers for `Choice`

@forki could you also please have a look at https://github.com/fsprojects/ProjectScaffold/pull/270 and https://github.com/fsprojects/FSharpx.Async/pull/31 ? others seem to be a bit passive. Thank you very much!

sideeffffect commented:
@tpetricek could you please have a look at this?

sideeffffect commented:
@forki @mausch @panesofglass @tpetricek 
it contains some useful helper functions that I would like to use from `FSharpx.Collections`
can we please merge this?

forki commented:
thx
